File: forgemind/registry.py
"""ForgekinRegistry — registry + lifecycle for forgekins.

Responsibilities:
- Register/unregister forgekins by id
- Lookup by id / name / type / capability
- Iterate active forgekins (energy > 0)
- Default process-wide registry (swappable for tests)
"""

from __future__ import annotations

from typing import Iterable

from flowforge.core.errors import ForgekinError
from flowforge.core.tracing import get_logger
from flowforge.forgemind.forgekin import Capability, Forgekin, ForgekinType

logger = get_logger("flowforge.forgemind.registry")


class ForgekinRegistry:
    """In-memory forgekin registry."""

    def __init__(self) -> None:
        self._by_id: dict[str, Forgekin] = {}

    def register(self, forgekin: Forgekin) -> None:
        if forgekin.forgekin_id in self._by_id:
            raise ForgekinError(f"Forgekin {forgekin.forgekin_id!r} already registered")
        self._by_id[forgekin.forgekin_id] = forgekin
        logger.info(
            f"registry: +forgekin id={forgekin.forgekin_id} name={forgekin.name!r} "
            f"type={forgekin.forgekin_type.value}"
        )

    def unregister(self, forgekin_id: str) -> Forgekin:
        if forgekin_id not in self._by_id:
            raise ForgekinError(f"Forgekin {forgekin_id!r} not found")
        fk = self._by_id.pop(forgekin_id)
        logger.info(f"registry: -forgekin id={forgekin_id}")
        return fk

    def get(self, forgekin_id: str) -> Forgekin:
        if forgekin_id not in self._by_id:
            raise ForgekinError(f"Forgekin {forgekin_id!r} not found")
        return self._by_id[forgekin_id]

    def find_by_name(self, name: str) -> list[Forgekin]:
        return [fk for fk in self._by_id.values() if fk.name == name]

    def find_by_type(self, forgekin_type: ForgekinType) -> list[Forgekin]:
        return [fk for fk in self._by_id.values() if fk.forgekin_type == forgekin_type]

    def find_by_capability(
        self,
        capability_name: str,
        min_proficiency: float = 0.5,
    ) -> list[Forgekin]:
        return [
            fk
            for fk in self._by_id.values()
            if fk.has_capability(capability_name, min_proficiency)
        ]

    def list_active(self) -> list[Forgekin]:
        return [fk for fk in self._by_id.values() if fk.state.energy > 0.0]

    def list_all(self) -> list[Forgekin]:
        return list(self._by_id.values())

    def count(self) -> int:
        return len(self._by_id)

    def select_owner(
        self,
        required_capabilities: list[str],
        exclude: Iterable[str] = (),
    ) -> Forgekin | None:
        """Pick the best forgekin to own a task requiring these capabilities.

        Selection heuristic: most matching capabilities × highest energy.
        Ties broken by registration order (deterministic for tests).
        """
        excluded = set(exclude)
        candidates: list[tuple[int, float, int, Forgekin]] = []
        for idx, fk in enumerate(self._by_id.values()):
            if fk.forgekin_id in excluded:
                continue
            if fk.state.energy <= 0.0:
                continue
            matched = sum(1 for c in required_capabilities if fk.has_capability(c))
            if matched == 0 and required_capabilities:
                continue
            min_prof = 1.0
            for c in required_capabilities:
                cap: Capability | None = fk.capabilities.get(c)
                if cap is not None:
                    min_prof = min(min_prof, cap.proficiency)
                else:
                    min_prof = 0.0
                    break
            candidates.append((matched, min_prof, -idx, fk))
        if not candidates:
            return None
        # Sort: most matched, then highest min proficiency, then earliest registered
        candidates.sort(key=lambda t: (t[0], t[1], t[2]), reverse=True)
        return candidates[0][3]

    def clear(self) -> None:
        self._by_id.clear()


# Process-wide default registry
_default_registry: ForgekinRegistry | None = None


def get_registry() -> ForgekinRegistry:
    global _default_registry
    if _default_registry is None:
        _default_registry = ForgekinRegistry()
    return _default_registry


def set_registry(registry: ForgekinRegistry | None) -> None:
    """Override default registry (used by tests for isolation)."""
    global _default_registry
    _default_registry = registry

File: packages/forgekin/species/src/registry.ts
/**
 * @flowforge/forgekin-species — ForgekinRegistry 注册表 + 生命周期
 *
 * TS 移植自 `forgemind/registry.py`：
 * - register/unregister by id / lookup by id/name/type/capability
 * - list_active（energy > 0）
 * - select_owner 启发式：最多匹配能力 × 最高最小熟练度 × 最早注册
 * - 默认注册表可替换（测试隔离）
 *
 * @module @flowforge/forgekin-species/registry
 */

import { Forgekin, ForgekinError, type Capability, type ForgekinType } from './models.js';

/** 内存态 Forgekin 注册表 */
export class ForgekinRegistry {
  /** id → forgekin（Map 保持插入顺序 = 注册顺序，对齐 Python dict） */
  private readonly byId = new Map<string, Forgekin>();

  register(forgekin: Forgekin): void {
    if (this.byId.has(forgekin.forgekinId)) {
      throw new ForgekinError(`Forgekin ${JSON.stringify(forgekin.forgekinId)} already registered`);
    }
    this.byId.set(forgekin.forgekinId, forgekin);
  }

  unregister(forgekinId: string): Forgekin {
    const fk = this.byId.get(forgekinId);
    if (fk === undefined) {
      throw new ForgekinError(`Forgekin ${JSON.stringify(forgekinId)} not found`);
    }
    this.byId.delete(forgekinId);
    return fk;
  }

  get(forgekinId: string): Forgekin {
    const fk = this.byId.get(forgekinId);
    if (fk === undefined) {
      throw new ForgekinError(`Forgekin ${JSON.stringify(forgekinId)} not found`);
    }
    return fk;
  }

  findByName(name: string): Forgekin[] {
    return [...this.byId.values()].filter((fk) => fk.name === name);
  }

  findByType(forgekinType: ForgekinType): Forgekin[] {
    return [...this.byId.values()].filter((fk) => fk.forgekinType === forgekinType);
  }

  findByCapability(capabilityName: string, minProficiency = 0.5): Forgekin[] {
    return [...this.byId.values()].filter((fk) => fk.hasCapability(capabilityName, minProficiency));
  }

  /** 活跃 Forgekin（energy > 0） */
  listActive(): Forgekin[] {
    return [...this.byId.values()].filter((fk) => fk.state.energy > 0.0);
  }

  listAll(): Forgekin[] {
    return [...this.byId.values()];
  }

  count(): number {
    return this.byId.size;
  }

  /**
   * 选择最适合承接任务的 Forgekin。
   *
   * 启发式（对齐 Python select_owner）：
   * 最多匹配能力 × 最高最小熟练度 × 最早注册（确定性，便于测试）。
   */
  selectOwner(requiredCapabilities: readonly string[], exclude: readonly string[] = []): Forgekin | null {
    const excluded = new Set(exclude);
    interface Candidate {
      matched: number;
      minProf: number;
      index: number;
      forgekin: Forgekin;
    }
    const candidates: Candidate[] = [];
    let idx = 0;
    for (const fk of this.byId.values()) {
      idx += 1;
      if (excluded.has(fk.forgekinId)) {
        continue;
      }
      if (fk.state.energy <= 0.0) {
        continue;
      }
      let matched = 0;
      for (const c of requiredCapabilities) {
        if (fk.hasCapability(c)) {
          matched += 1;
        }
      }
      if (matched === 0 && requiredCapabilities.length > 0) {
        continue;
      }
      let minProf = 1.0;
      for (const c of requiredCapabilities) {
        const cap: Capability | undefined = fk.capabilities.get(c);
        if (cap !== undefined) {
          minProf = Math.min(minProf, cap.proficiency);
        } else {
          minProf = 0.0;
          break;
        }
      }
      candidates.push({ matched, minProf, index: idx, forgekin: fk });
    }
    if (candidates.length === 0) {
      return null;
    }
    // 排序：matched 降序 → minProf 降序 → index 升序（最早注册优先）
    candidates.sort((a, b) =>
      a.matched !== b.matched
        ? b.matched - a.matched
        : a.minProf !== b.minProf
          ? b.minProf - a.minProf
          : a.index - b.index,
    );
    return candidates[0]?.forgekin ?? null;
  }

  clear(): void {
    this.byId.clear();
  }
}

// ── 默认注册表（进程级，可替换用于测试隔离）──────────────────────

let defaultRegistry: ForgekinRegistry | null = null;

/** 获取进程级默认注册表（懒创建） */
export function getRegistry(): ForgekinRegistry {
  if (defaultRegistry === null) {
    defaultRegistry = new ForgekinRegistry();
  }
  return defaultRegistry;
}

/** 覆盖默认注册表（传 null 重置；测试隔离用） */
export function setRegistry(registry: ForgekinRegistry | null): void {
  defaultRegistry = registry;
}
